components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java

```diff
@@ -128,476 +131,628 @@ public void init() throws IdentityOAuth2Exception {
-          attribute of a suitable <SubjectConfirmationData> element.
-         */
+    private void setValuesInMessageContext(OAuthTokenReqMessageContext tokReqMsgCtx, Assertion assertion,
+                                           IdentityProvider identityProvider, String tenantDomain)
+            throws IdentityOAuth2Exception {
+        setUserInMessageContext(tokReqMsgCtx, identityProvider, assertion, tenantDomain);
+        tokReqMsgCtx.setScope(tokReqMsgCtx.getOauth2AccessTokenReqDTO().getScope());
+        // Storing the Assertion. This will be used in OpenID Connect for example
+        tokReqMsgCtx.addProperty(OAuthConstants.OAUTH_SAML2_ASSERTION, assertion);
+    }
 
-        /*
-          The <Subject> element MUST contain at least one <SubjectConfirmation> element that allows the
-          authorization server to confirm it as a Bearer Assertion.  Such a <SubjectConfirmation> element MUST
-          have a Method attribute with a value of "urn:oasis:names:tc:SAML:2.0:cm:bearer".  The
-          <SubjectConfirmation> element MUST contain a <SubjectConfirmationData> element, unless the Assertion
-          has a suitable NotOnOrAfter attribute on the <Conditions> element, in which case the
-          <SubjectConfirmationData> element MAY be omitted. When present, the <SubjectConfirmationData> element
-          MUST have a Recipient attribute with a value indicating the token endpoint URL of the authorization
-          server (or an acceptable alias).  The authorization server MUST verify that the value of the Recipient
-          attribute matches the token endpoint URL (or an acceptable alias) to which the Assertion was delivered.
-          The <SubjectConfirmationData> element MUST have a NotOnOrAfter attribute that limits the window during
-          which the Assertion can be confirmed.  The <SubjectConfirmationData> element MAY also contain an Address
-          attribute limiting the client address from which the Assertion can be delivered.  Verification of the
-          Address is at the discretion of the authorization server.
-         */
+    private void invokeExtension(OAuthTokenReqMessageContext tokReqMsgCtx) throws IdentityOAuth2Exception {
+        // Invoking extension
+        SAML2TokenCallbackHandler callback = OAuthServerConfiguration.getInstance().getSAML2TokenCallbackHandler();
+        if (callback != null) {
+            if (log.isDebugEnabled()){
+                log.debug("Invoking the SAML2 Token callback handler");
+            }
+            callback.handleSAML2Token(tokReqMsgCtx);
+        }
+    }
 
-        DateTime notOnOrAfterFromConditions = null;
-        Map<DateTime,DateTime> notOnOrAfterFromAndNotBeforeSubjectConfirmations = new HashMap<>();
-        boolean bearerFound = false;
-        List<String> recipientURLS = new ArrayList<>();
+    private void validateSignatureAgainstIdpCertificate(Assertion assertion, String tenantDomain,
```

Comment:
Shall we make this protected so that anyone who wants to plugin there own keystore/validation mechanism can override this alone?